File: src/server/controllers/wow-config.ts
import type { Router } from 'express';

import { clearProjectCacheDir, getProjectCacheDirSize } from '@/lib/wow/clear-project-cache';
import { normalizeInstallDirectory } from '@/lib/wow/normalize-install-directory';
import {
  applyWowConfig,
  discoverLocalBuilds,
  discoverRemoteBuilds,
  getWowConfigStatus,
  resetWowConfig,
} from '@/lib/wow/wow-config-service';
import type { WowConfig } from '@/lib/wow/wow-config-state';

import { isSharedHosting } from '../config';
import { assertDesktopOnly, desktopOnlyStatus } from '../shared-hosting';
import { pickNativeFolder } from '../utils/pick-folder';

const WOW_CONFIG_SHARED_HOSTING_LOCKED = 'WoW installation cannot be changed in shared hosting mode.';

function assertWowConfigMutable(): void {
  if (isSharedHosting) {
    throw new Error(WOW_CONFIG_SHARED_HOSTING_LOCKED);
  }
}

function wowConfigErrorStatus(error: Error): number {
  return error.message === WOW_CONFIG_SHARED_HOSTING_LOCKED ? 403 : 400;
}

export function ControllerWowConfig(router: Router): void {
  router.get('/wow-config/status', async (_req, res, next) => {
    try {
      res.setHeader('Cache-Control', 'no-store');
      res.json(await getWowConfigStatus());
    } catch (e) {
      next(e);
    }
  });

  router.post('/wow-config/pick-local-folder', (_req, res) => {
    try {
      assertWowConfigMutable();
      const startDirectory = typeof _req.body?.installDirectory === 'string'
        ? normalizeInstallDirectory(_req.body.installDirectory)
        : undefined;
      const installDirectory = pickNativeFolder(
        'Select World of Warcraft install folder',
        startDirectory,
      );
      if (!installDirectory) {
        res.json({ cancelled: true });
        return;
      }
      res.json({ installDirectory });
    } catch (e) {
      const err = e as Error;
      res.status(wowConfigErrorStatus(err)).json({ error: err.message });
    }
  });

  router.post('/wow-config/reset', async (_req, res) => {
    try {
      assertWowConfigMutable();
      await resetWowConfig();
      res.json(await getWowConfigStatus());
    } catch (e) {
      const err = e as Error;
      res.status(wowConfigErrorStatus(err)).json({ error: err.message });
    }
  });

  router.get('/wow-config/cache-size', async (_req, res, next) => {
    try {
      assertDesktopOnly();
      const bytes = await getProjectCacheDirSize();
      res.json({ bytes });
    } catch (e) {
      const err = e as Error;
      if (desktopOnlyStatus(err) === 403) {
        res.status(403).json({ error: err.message });
        return;
      }
      next(e);
    }
  });

  router.post('/wow-config/clear-cache', async (_req, res) => {
    try {
      assertDesktopOnly();
      await clearProjectCacheDir();
      res.json(await getWowConfigStatus());
    } catch (e) {
      const err = e as Error;
      res.status(desktopOnlyStatus(err)).json({ error: err.message });
    }
  });

  router.post('/wow-config/discover-local', async (req, res) => {
    try {
      assertWowConfigMutable();
      const installDirectory = req.body?.installDirectory;
      if (typeof installDirectory !== 'string' || !installDirectory.trim()) {
        res.status(400).json({ error: 'installDirectory is required' });
        return;
      }
      const builds = await discoverLocalBuilds(normalizeInstallDirectory(installDirectory));
      res.json({ builds });
    } catch (e) {
      const err = e as Error;
      res.status(wowConfigErrorStatus(err)).json({ error: err.message });
    }
  });

  router.post('/wow-config/discover-remote', async (req, res) => {
    try {
      assertWowConfigMutable();
      const regionTag = req.body?.regionTag;
      if (typeof regionTag !== 'string' || !regionTag.trim()) {
        res.status(400).json({ error: 'regionTag is required' });
        return;
      }
      const builds = await discoverRemoteBuilds(regionTag.trim());
      res.json({ builds });
    } catch (e) {
      const err = e as Error;
      res.status(wowConfigErrorStatus(err)).json({ error: err.message });
    }
  });

  router.post('/wow-config/apply', async (req, res) => {
    try {
      assertWowConfigMutable();
      const {
        mode, installDirectory, regionTag, product,
      } = req.body ?? {};
      if (mode !== 'local' && mode !== 'remote') {
        res.status(400).json({ error: 'mode must be "local" or "remote"' });
        return;
      }
      if (typeof product !== 'string' || !product.trim()) {
        res.status(400).json({ error: 'product is required' });
        return;
      }

      let config: WowConfig;
      if (mode === 'local') {
        if (typeof installDirectory !== 'string' || !installDirectory.trim()) {
          res.status(400).json({ error: 'installDirectory is required for local mode' });
          return;
        }
        config = {
          mode: 'local',
          installDirectory: normalizeInstallDirectory(installDirectory),
          product: product.trim(),
        };
      } else {
        if (typeof regionTag !== 'string' || !regionTag.trim()) {
          res.status(400).json({ error: 'regionTag is required for remote mode' });
          return;
        }
        config = {
          mode: 'remote',
          regionTag: regionTag.trim(),
          product: product.trim(),
        };
      }

      const cascInfo = await applyWowConfig(config);
      const status = await getWowConfigStatus();
      res.json({ cascInfo, status });
    } catch (e) {
      const err = e as Error;
      res.status(wowConfigErrorStatus(err)).json({ error: err.message });
    }
  });
}

File: golang/internal/server/api/wow_config.go
package api

import (
	"net/http"
	"strings"

	"github.com/pqhuy98/wow-converter/internal/server/pickfolder"
	"github.com/pqhuy98/wow-converter/internal/workspace"
	"github.com/pqhuy98/wow-converter/internal/wow/wowconfig"
)

func registerWowConfig(r Router, d *Deps) {
	svc := wowconfig.NewService(d.Client)

	r.Get("/wow-config/status", func(w http.ResponseWriter, req *http.Request) {
		setNoStore(w)
		sendJSON(w, http.StatusOK, svc.GetStatus(req.Context()))
	})

	r.Get("/wow-config/cache-size", func(w http.ResponseWriter, _ *http.Request) {
		if err := assertDesktopOnly(d.Config.IsSharedHosting); err != nil {
			sendError(w, http.StatusForbidden, err.Error())
			return
		}
		bytes, err := workspace.ProjectCacheDirSize()
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		sendJSON(w, http.StatusOK, map[string]any{"bytes": bytes})
	})

	r.Post("/wow-config/pick-local-folder", func(w http.ResponseWriter, req *http.Request) {
		if err := wowconfig.AssertMutable(d.Config.IsSharedHosting); err != nil {
			status := http.StatusBadRequest
			if err.Error() == wowconfig.SharedHostingLockedMessage() {
				status = http.StatusForbidden
			}
			sendError(w, status, err.Error())
			return
		}
		body, err := readJSONMap(req)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		startDirectory := ""
		if raw, ok := body["installDirectory"].(string); ok {
			startDirectory = wowconfig.NormalizeInstallDirectory(raw)
		}
		installDirectory, err := pickfolder.PickNativeFolder(
			"Select World of Warcraft install folder",
			startDirectory,
		)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		if installDirectory == "" {
			sendJSON(w, http.StatusOK, map[string]any{"cancelled": true})
			return
		}
		sendJSON(w, http.StatusOK, map[string]any{"installDirectory": wowconfig.NormalizeInstallDirectory(installDirectory)})
	})

	r.Post("/wow-config/reset", func(w http.ResponseWriter, req *http.Request) {
		if err := wowconfig.AssertMutable(d.Config.IsSharedHosting); err != nil {
			status := http.StatusBadRequest
			if err.Error() == wowconfig.SharedHostingLockedMessage() {
				status = http.StatusForbidden
			}
			sendError(w, status, err.Error())
			return
		}
		if err := svc.Reset(req.Context()); err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		sendJSON(w, http.StatusOK, svc.GetStatus(req.Context()))
	})

	r.Post("/wow-config/clear-cache", func(w http.ResponseWriter, req *http.Request) {
		if err := assertDesktopOnly(d.Config.IsSharedHosting); err != nil {
			sendError(w, http.StatusForbidden, err.Error())
			return
		}
		if err := svc.ClearCache(req.Context()); err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		sendJSON(w, http.StatusOK, svc.GetStatus(req.Context()))
	})

	r.Post("/wow-config/discover-local", func(w http.ResponseWriter, req *http.Request) {
		if err := wowconfig.AssertMutable(d.Config.IsSharedHosting); err != nil {
			status := http.StatusBadRequest
			if err.Error() == wowconfig.SharedHostingLockedMessage() {
				status = http.StatusForbidden
			}
			sendError(w, status, err.Error())
			return
		}
		body, err := readJSONMap(req)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		installDirectory, _ := body["installDirectory"].(string)
		installDirectory = wowconfig.NormalizeInstallDirectory(installDirectory)
		if installDirectory == "" {
			sendError(w, http.StatusBadRequest, "installDirectory is required")
			return
		}
		builds, err := svc.DiscoverLocalBuilds(req.Context(), installDirectory)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		sendJSON(w, http.StatusOK, map[string]any{"builds": builds})
	})

	r.Post("/wow-config/discover-remote", func(w http.ResponseWriter, req *http.Request) {
		if err := wowconfig.AssertMutable(d.Config.IsSharedHosting); err != nil {
			status := http.StatusBadRequest
			if err.Error() == wowconfig.SharedHostingLockedMessage() {
				status = http.StatusForbidden
			}
			sendError(w, status, err.Error())
			return
		}
		body, err := readJSONMap(req)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		regionTag, _ := body["regionTag"].(string)
		regionTag = strings.TrimSpace(regionTag)
		if regionTag == "" {
			sendError(w, http.StatusBadRequest, "regionTag is required")
			return
		}
		builds, err := svc.DiscoverRemoteBuilds(req.Context(), regionTag)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		sendJSON(w, http.StatusOK, map[string]any{"builds": builds})
	})

	r.Post("/wow-config/apply", func(w http.ResponseWriter, req *http.Request) {
		if err := wowconfig.AssertMutable(d.Config.IsSharedHosting); err != nil {
			status := http.StatusBadRequest
			if err.Error() == wowconfig.SharedHostingLockedMessage() {
				status = http.StatusForbidden
			}
			sendError(w, status, err.Error())
			return
		}
		body, err := readJSONMap(req)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		mode, _ := body["mode"].(string)
		product, _ := body["product"].(string)
		product = strings.TrimSpace(product)
		if mode != "local" && mode != "remote" {
			sendError(w, http.StatusBadRequest, `mode must be "local" or "remote"`)
			return
		}
		if product == "" {
			sendError(w, http.StatusBadRequest, "product is required")
			return
		}

		cfg := wowconfig.Config{Mode: wowconfig.Mode(mode), Product: product}
		if mode == "local" {
			installDirectory, _ := body["installDirectory"].(string)
			installDirectory = wowconfig.NormalizeInstallDirectory(installDirectory)
			if installDirectory == "" {
				sendError(w, http.StatusBadRequest, "installDirectory is required for local mode")
				return
			}
			cfg.InstallDirectory = installDirectory
		} else {
			regionTag, _ := body["regionTag"].(string)
			regionTag = strings.TrimSpace(regionTag)
			if regionTag == "" {
				sendError(w, http.StatusBadRequest, "regionTag is required for remote mode")
				return
			}
			cfg.RegionTag = regionTag
		}

		cascInfo, err := svc.Apply(req.Context(), cfg, true)
		if err != nil {
			sendError(w, http.StatusBadRequest, err.Error())
			return
		}
		sendJSON(w, http.StatusOK, map[string]any{
			"cascInfo": cascInfo,
			"status":   svc.GetStatus(req.Context()),
		})
	})
}
